utils/fifoqueue/fifoqueue.go

```diff
@@ -108,5 +122,19 @@ func (q *FifoQueue) Pop() (interface{}, bool) {
 
 // Len returns the current length of the queue.
 func (q *FifoQueue) Len() int {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
 	return q.queue.Len()
 }
+
+func (q *FifoQueue) Put(msg *engine.Message) bool {
```

Comment:
Seems odd that general purpose component implements interface from engine. I would like to keep it without extra dependencies.

Reply:
👍  I was thinking we could either add an `engine.Message`-typed implementation wrapping this utility in `engine` or move the `fifoqueue` into the `engine` directory if this is the only place it is used

Reply:
wrapping it in `engine` is a bit complex, because I would still need to import `utils.fifoqueue` to use the constructor and the constructor options. Then the type and its constructor would be from different packages.

Yeah, I would move to engine directly to keep `engine/common/fifoqueue` to be simple